pinot-broker/src/main/java/com/linkedin/pinot/broker/routing/builder/PartitionAwareRealtimeRoutingTableBuilder.java

```diff
@@ -0,0 +1,150 @@
+/**
+ * Copyright (C) 2014-2016 LinkedIn Corp. (pinot-core@linkedin.com)
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *         http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.linkedin.pinot.broker.routing.builder;
+
+import com.linkedin.pinot.common.metadata.segment.SegmentZKMetadata;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.Set;
+import java.util.SortedSet;
+
+import org.apache.helix.model.ExternalView;
+import org.apache.helix.model.InstanceConfig;
+
+import com.linkedin.pinot.common.utils.CommonConstants;
+import com.linkedin.pinot.common.utils.LLCSegmentName;
+import com.linkedin.pinot.common.utils.SegmentName;
+import com.linkedin.pinot.common.metadata.ZKMetadataProvider;
+import com.linkedin.pinot.common.response.ServerInstance;
+import com.linkedin.pinot.transport.common.SegmentId;
+
+/**
+ * Partition aware routing table builder for the Kafka low level consumer.
+ *
+ * In contrast to the offline partition aware routing builder where the replica group aware segment assignment can be
+ * assumed, we do not use the concept of replica group for a realtime table. The routing look up table is simply built
+ * from the external view.
+ *
+ */
+public class PartitionAwareRealtimeRoutingTableBuilder extends AbstractPartitionAwareRoutingTableBuilder {
+
+  @Override
+  public void computeRoutingTableFromExternalView(String tableName, ExternalView externalView,
+      List<InstanceConfig> instanceConfigList) {
+
+    // TODO: enforce the replica group concept when possible
+    //
+    // Kafka low level does not support for the concept of replica group. However, it assigns servers sequentially
+    // when building the kafka partition table. Let's assume that we have 4 servers, 2 replicas, 4 partitions.
+    //
+    // partition | replica1 | replica2
+    //         0 |  server0 |  server1
+    //         1 |  server2 |  server3
+    //         2 |  server0 |  server1
+    //         3 |  server2 |  server3
+    //
+    // When there is no partitioning on data, we can still reduce the number of servers to route if we pick servers
```

Comment:
No assumption should be made about how the partitions will be assigned across servers. The only assumption that can be made is that the number of replicas will be honored. In fact, this serial assignment that has been pointed out in the comments is actually because helix code had a bug https://issues.apache.org/jira/browse/HELIX-631 Once this is resolved and we start using this call, the Kafka partition assignment will be totally opaque to us.